mac/Keyman4MacIM/Keyman4MacIM/TextApiCompliance.m

```diff
@@ -76,77 +76,116 @@ -(NSString *)description
-      self.hasCompliantSelectionApi = NO;
-      self.complianceUncertain = NO;
-      [self.appDelegate logDebugMessage:@"TextApiCompliance testComplianceAfterInsert certain, non-compliant, range is NSNotFound"];
-    } else if (selectionRange.location == 0) {
-      // NO for certain, after an insertText we cannot be at location 0
-      self.hasCompliantSelectionApi = NO;
-      self.complianceUncertain = NO;
-      [self.appDelegate logDebugMessage:@"TextApiCompliance testComplianceAfterInsert certain, non-compliant, location = 0"];
-    } else if (selectionRange.location > 0) {
-      if (NSEqualRanges(selectionRange, self.initialSelection)) {
-        // NO for certain, when the selection is unchanged after an insert
-        self.hasCompliantSelectionApi = NO;
-        self.complianceUncertain = NO;
-        [self.appDelegate logDebugMessage:@"TextApiCompliance testComplianceAfterInsert certain, non-compliant, selection the same: initial location = %@, new location = %@", NSStringFromRange(self.initialSelection), NSStringFromRange(selectionRange)];
-      } else {
-        // YES for certain, when the selection is changed after an insert
-        self.hasCompliantSelectionApi = YES;
-        self.complianceUncertain = NO;
-        [self.appDelegate logDebugMessage:@"TextApiCompliance testComplianceAfterInsert certain, compliant, selection changed from initial location = %@ to new location = %@", NSStringFromRange(self.initialSelection), NSStringFromRange(selectionRange)];
-      }
-    }
-    
-    [self.appDelegate logDebugMessage:@"TextApiCompliance testComplianceAfterInsert, self.hasWorkingSelectionApi = %@ for app %@", self.hasCompliantSelectionApi?@"yes":@"no", self.clientApplicationId];
-    [self.appDelegate logDebugMessage:@"TextApiCompliance testComplianceAfterInsert TextApiCompliance: %@", self];
- } else {
-   [self.appDelegate logDebugMessage:@"TextApiCompliance testSelectionApiAfterInsert, compliance is already known"];
+/**
+ * If complianceUncertain is true, checking the location after an insert can confirm whether the app is compliant.
+ * Delete and insert are what core instructed us to do.
+ * Text that was selected in the client when the key was processed is irrelevant as it does not affect the location.
+ */
+-(void) checkComplianceAfterInsert:(id) client delete:(NSString *)textToDelete insert:(NSString *)textToInsert {
+  
+  // return if compliance is already certain
+  if(!self.complianceUncertain) {
+    [self.appDelegate logDebugMessage:@"TextApiCompliance testSelectionApiAfterInsert, compliance is already known"];
+    return;
+  }
+  
+  NSRange selectionRange = [client selectedRange];
+  if ([self validateNewLocation:selectionRange.location delete:textToDelete]) {
+    BOOL changeExpected = [self isLocationChangeExpectedOnInsert:textToDelete insert:textToInsert];
+    BOOL locationChanged = [self hasLocationChanged:selectionRange];
```

Comment:
As far as I can see, we're not double-checking the degree to which the selection range has shifted?  If a keystroke happened to emit two characters, I don't see where this validates that the selection range shifted by that amount.

Granted, I'm new to all this compliance-checking for Mac stuff, so maybe that's not a concern?  Admittedly, I only reviewed the new lines rather than try to also make sense of the old form too.

Given what this PR sets out to do, I'll say that the part that double-checks that the selection-range index should have a net move does seem to be correct to me, at least under the assumption of range width 0.  If there's pre-selected text, that would probably complicate things further.

Reply:
Based on all the apps that I've tested, there is no need to check for the amount that the location has changed. I was thinking of doing that, but there does not seem to be anything to gain in doing so. The invalid responses are either NSNotFound or some other constant for the location. 

The exception was one app which returned an invalid location that was changing by the expected amount with each keystroke. Because the intial location was incorrect the correct delta did not make it compliant, so I had to hard-code that as non-compliant. (In that case, it would have still appeared to be compliant even with the check you are suggesting.)